src/Microsoft.Azure.SignalR.Common/ServiceConnections/ServiceConnectionBase.cs

```diff
@@ -30,6 +31,11 @@ internal abstract class ServiceConnectionBase : IServiceConnection
         private readonly ReadOnlyMemory<byte> _cachedPingBytes;
         private readonly HandshakeRequestMessage _handshakeRequest;
 
+        private readonly SemaphoreSlim _closeConnectionLock = new SemaphoreSlim(1, 1);
+
+        protected volatile Task _incomingTask = Task.CompletedTask;
```

Comment:
Is there any change these could be assigned *after* shutting down?

Reply:
Sadly it would happen.

Considering if a connection hadn't started (running `StartAsync`) while the matching `CloseAsync` was being called. The `WriteAsync` would throw an exception and we have no chance to gracefully stop the `incoming` Task. 

But the odds are very low.

If we want to fully prevent this from happening, we have to introduce a lock to avoid the `StartAsync` and `CloseAsync` run at the same time, which could make our codes even more complex.

Should we do it?

Reply:
You don’t have to. There are other things that could be done:
1. Do notching, assume it’s rare enough and leave it. I’d the process is being torn down it’s not a big deal but if not, we could leak the tasks.

2. Trigger a cancellation token in close that process input would respect. It means that even if the ProcessIncoming started, it would eventually be cancelled by the token.

3. Create a TaskCompletionSource and trigger it based on the completion of ProcessIncoming and ProcessOutgoing. This way you always have the same reference which just changes state.

Reply:
I'll control the process by using `InterLock.Exchange`, I've already learned this is an equivalent of `CAS` in C#.